src/package_manifest.rs

```diff
@@ -88,78 +88,75 @@ impl AsRef<PackageManifest> for PackageManifest {
-
-        get_dependency_group("dependencies")
+        static DEPENDENCY_GROUPS: &[&str] = &[
+            "dependencies",
+            "devDependencies",
+            "optionalDependencies",
+            "peerDependencies",
+        ];
+
+        DEPENDENCY_GROUPS
             .iter()
-            .chain(get_dependency_group("devDependencies").iter())
-            .chain(get_dependency_group("optionalDependencies").iter())
-            .chain(get_dependency_group("peerDependencies").iter())
+            // only iterate over the objects corresponding to each dependency group
+            .filter_map(|dependency_group| {
+                self.contents.extra_fields.get(dependency_group)?.as_object()
+            })
+            // get all dependency names from all groups
+            .flat_map(|dependency_group_value| dependency_group_value.keys())
             // filter out external packages
             .filter_map(|package_name| {
                 package_manifests_by_package_name
-                    .get(*package_name)
-                    .cloned()
-                    .map(|thing| thing.as_ref())
+                    .get(package_name)
+                    .map(|&thing| thing.as_ref())
             })
-            .collect()
     }
 
     pub fn transitive_internal_dependency_package_names<'a>(
         &self,
         package_manifest_by_package_name: &HashMap<String, &'a PackageManifest>,
     ) -> Vec<&'a PackageManifest> {
         // Depth-first search all transitive internal dependencies of package
-        let mut seen_package_names: HashSet<&str> = HashSet::new();
-        let mut internal_dependencies: HashSet<String> = HashSet::new();
-        let mut to_visit_package_manifests: VecDeque<&PackageManifest> = VecDeque::new();
+        let mut seen_package_names = HashSet::new();
+        let mut internal_dependencies = HashSet::new();
+        let mut to_visit_package_manifests = VecDeque::new();
 
         to_visit_package_manifests.push_back(self);
 
-        while !to_visit_package_manifests.is_empty() {
-            let current_manifest = to_visit_package_manifests.pop_front().unwrap();
+        while let Some(current_manifest) = to_visit_package_manifests.pop_front() {
             seen_package_names.insert(&current_manifest.contents.name);
 
             for dependency in current_manifest
-                .get_internal_dependencies(package_manifest_by_package_name)
-                .iter()
+                .internal_dependencies_iter(package_manifest_by_package_name)
             {
                 internal_dependencies.insert(dependency.contents.name.to_owned());
-                if !seen_package_names.contains(&dependency.contents.name.as_ref()) {
+                if !seen_package_names.contains(&dependency.contents.name) {
                     to_visit_package_manifests.push_back(dependency);
                 }
             }
         }
 
         internal_dependencies
-            .iter()
+            .into_iter()
```

Comment:
And then one more time, can  you explain the reasoning behind how this change is beneficial? 

I figured the less ownership you'd exert over data the better, because using references fosters easier extension by new code in the future

Reply:
This is true, using `iter` would allow us to not consider this instance of ownership if this function were extended, as `internal_dependencies` wouldn't be consumed. I don't know that I'd consider this a net positive in general though: ownership allows us to specify when things should "die". It is our intention that at the end of the function as it is defined now, `internal_dependencies` has been completely used. 

Requiring that code added to the end of this same function and using `internal_dependencies` reconcile this ownership is a signal to the authors of that code: "you are not using `internal_dependencies` as originally intended, it was made to die here."

All of this said, because `internal_dependencies` has very little "in progress" information and the function up to this point is essentially finished with it, there isn't *harm* in only using it by reference. My personal preference is still to consume where practical, though.

Reply:
It's also a hill I won't die on ;)

Reply:
Removed it[](url)

Reply:
Oh no! I apologize for the slower-than-usual reviews, I was slow on the account-switching draw.

I like the diff, is there a reason you changed your mind? I appreciate you writing out that philosophical approach, this is foreign for me as a new participant in the borrow-checking arena. Your description makes a lot of sense, it's an idiom that will need more practice to become second nature